repo/content/content_manager_indexes.go

```diff
@@ -0,0 +1,380 @@
+
+	// generate random log entry ID to ensure that when it is hashed later,
+	// we don't get any collisions.
+	logEntry.RandomID = make([]byte, 32)
+	if _, err := rand.Read(logEntry.RandomID); err != nil {
+		return errors.Wrap(err, "unable to get random log entry ID")
+	}
+
+	for _, indexBlob := range indexBlobs {
+		if err := bm.addIndexBlobsToBuilder(ctx, bld, indexBlob, opt); err != nil {
+			return errors.Wrap(err, "error adding index to builder")
+		}
+
+		logEntry.InputBlobs = append(logEntry.InputBlobs, indexBlob.BlobID)
+	}
+
+	var buf bytes.Buffer
+	if err := bld.Build(&buf); err != nil {
+		return errors.Wrap(err, "unable to build an index")
+	}
+
+	compactedIndexBlob, err := bm.writePackIndexesNew(ctx, buf.Bytes())
+	if err != nil {
+		return errors.Wrap(err, "unable to write compacted indexes")
+	}
+
+	// compaction wrote index blob that's the same as one of the sources
+	// it must be a no-op.
+	for _, indexBlob := range indexBlobs {
+		if indexBlob.BlobID == compactedIndexBlob {
+			formatLog(ctx).Debugf("compaction was a no-op")
+			return nil
+		}
+	}
+
+	logEntry.OutputBlobs = append(logEntry.OutputBlobs, compactedIndexBlob)
+
+	logEntryBytes, err := json.Marshal(logEntry)
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal log entry bytes")
+	}
+
+	compactionLogBlobID, err := bm.encryptAndWriteBlobNotLocked(ctx, logEntryBytes, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "unable to write compaction log")
+	}
+
+	formatLog(ctx).Debugf("compacted indexes %v into %v and wrote log %v in %v", logEntry.InputBlobs, logEntry.OutputBlobs, compactionLogBlobID, time.Since(t0)) // allow:no-inject-time
+
+	if err := bm.deleteOldIndexBlobs(ctx, compactionLogBlobID); err != nil {
+		return errors.Wrap(err, "error deleting old index blobs")
+	}
+
+	return nil
+}
+
+func (bm *lockFreeManager) getCompactionLogEntries(ctx context.Context, blobs []blob.Metadata) (map[blob.ID]*compactionLogEntry, error) {
+	results := map[blob.ID]*compactionLogEntry{}
+
+	for _, cb := range blobs {
+		data, err := bm.getIndexBlobInternal(ctx, cb.BlobID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to read compaction blob %q", cb.BlobID)
+		}
+
+		le := &compactionLogEntry{}
+
+		if err := json.Unmarshal(data, le); err != nil {
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (bm *Manager) deleteOldIndexBlobs(ctx context.Context, latestBlobID blob.ID) error {
+	allCompactionLogBlobs, err := bm.listCache.listIndexBlobs(ctx, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing compaction log blobs")
+	}
+
+	// look for server-assigned timestamp of the compaction log entry we just wrote as a reference.
+	// we're assuming server-generated timestamps are somewhat reasonable and time is moving
+	latestBlobTime := blobTime(allCompactionLogBlobs, latestBlobID)
```

Comment:
Is this really the server-side timestamp? Wouldn't the metadata for `latesBlobID` come from the `ownWritesCache` and thus have a local timestamp?

Reply:
I'm going to add support for server-side timestamps in #440 before moving forward with this PR.

Reply:
done, own writes cache now supports server-side timestamps. PTAL